src/compiler/types/validate-primary-package-output-target.ts

```diff
@@ -0,0 +1,229 @@
+import { buildWarn, isString, isValidatableOutputTarget, normalizePath } from '@utils';
+import { join, relative } from 'path';
+
+import type * as d from '../../declarations';
+import { packageJsonError, packageJsonWarn } from './package-json-log.utils';
+
+type PrimaryPackageOutputTargetRecommendedConfig = {
+  getModulePath?: (rootDir: string, outputTargetDir: string) => string | null;
+  getTypesPath?: (rootDir: string, outputTargetConfig: any) => string | null;
+};
+
+const PRIMARY_PACKAGE_TARGET_CONFIGS: {
+  [key: string]: PrimaryPackageOutputTargetRecommendedConfig;
+} = {
+  dist: {
+    getModulePath: (rootDir: string, outputTargetDir: string) =>
+      normalizePath(relative(rootDir, join(outputTargetDir, 'index.js'))),
+    getTypesPath: (rootDir: string, outputTargetConfig: d.OutputTargetDist) =>
+      normalizePath(relative(rootDir, join(outputTargetConfig.typesDir, 'index.d.ts'))),
+  },
+  'dist-collection': {
+    getModulePath: (rootDir: string, outputTargetDir: string) =>
+      normalizePath(relative(rootDir, join(outputTargetDir, 'index.js'))),
+  },
+  'dist-custom-elements': {
+    getModulePath: (rootDir: string, outputTargetDir: string) =>
+      normalizePath(relative(rootDir, join(outputTargetDir, 'index.js'))),
+    getTypesPath: (rootDir: string, outputTargetConfig: d.OutputTargetDistCustomElements) => {
+      return outputTargetConfig.generateTypeDeclarations
+        ? normalizePath(relative(rootDir, join(rootDir, 'dist/types/index.d.ts')))
+        : null;
+    },
+  },
+  'dist-types': {
+    getTypesPath: (rootDir: string, outputTargetConfig: d.OutputTargetDistTypes) =>
+      normalizePath(relative(rootDir, join(outputTargetConfig.typesDir, 'index.d.ts'))),
+  },
+};
```

Comment:
you might get nicer typing here if you use my new absolute favorite thing in typescript, the `satisfies` keyword:

```suggestion
const PRIMARY_PACKAGE_TARGET_CONFIGS = {
  dist: {
    getModulePath: (rootDir: string, outputTargetDir: string) =>
      normalizePath(relative(rootDir, join(outputTargetDir, 'index.js'))),
    getTypesPath: (rootDir: string, outputTargetConfig: d.OutputTargetDist) =>
      normalizePath(relative(rootDir, join(outputTargetConfig.typesDir, 'index.d.ts'))),
  },
  'dist-collection': {
    getModulePath: (rootDir: string, outputTargetDir: string) =>
      normalizePath(relative(rootDir, join(outputTargetDir, 'index.js'))),
  },
  'dist-custom-elements': {
    getModulePath: (rootDir: string, outputTargetDir: string) =>
      normalizePath(relative(rootDir, join(outputTargetDir, 'index.js'))),
    getTypesPath: (rootDir: string, outputTargetConfig: d.OutputTargetDistCustomElements) => {
      return outputTargetConfig.generateTypeDeclarations
        ? normalizePath(relative(rootDir, join(rootDir, 'dist/types/index.d.ts')))
        : null;
    },
  },
  'dist-types': {
    getTypesPath: (rootDir: string, outputTargetConfig: d.OutputTargetDistTypes) =>
      normalizePath(relative(rootDir, join(outputTargetConfig.typesDir, 'index.d.ts'))),
  },
} satisfies Record<string, PrimaryPackageOutputTargetRecommendedConfig>;
```

then you'll still get type-checking that `getTypesPath` and `getModulePath` have the right type (if they're defined) but you'll _also_ still get a more specific type for the `PRIMARY_PACKAGE_TARGET_CONFIGS` than a `Record<...>`

Reply:
Oooh yeah this does play nicely with your first suggestion to get type-safety on the keys as well (which is something I was having trouble with)